### ISSUE TITLE
cov_params_XYZ naming conventions in tsa

### ISSUE THREAD

**Issue by jbrockmendel** [statsmodels/statsmodels#3675]
There are a bunch of similar `cov_params_XYZ` methods in `tsa.statespace.mlemodel` and `tsa.regime_switching.markov_switching`.  If some small mismatches in naming conventions were resolved, a bunch of the logic could be generalized.

Both [mlemodel.MLEResults](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/tsa/statespace/mlemodel.py#L1611) and [markov_switching.MarkovSwitchingResults](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/tsa/regime_switching/markov_switching.py#L1918) have a `_get_robustcov_results` method that dispatches to one of many options based on the `cov_type` kwarg. 

In `mlemodel`, if `cov_type="robust_approx"`, it gets passed through to `cov_params_robust`, which I find counter-intuitive given that there exists a `cov_params_robust_approx` method.

I'll add a couple other smaller cases if there's an interest.  What do you think about aligning the keyword usage between these two?

**Comment by ChadFulton** [comment]
> In mlemodel, if cov_type="robust_approx", it gets passed through to cov_params_robust, which I find counter-intuitive given that there exists a cov_params_robust_approx method.

This looks like a bug to me - it looks like there's no way to get the `cov_params_robust_approx`. Thanks for pointing that out.

> What do you think about aligning the keyword usage between these two?

This sounds good, but from a quick look I couldn't quite tell what you meant by keyword mismatches. The main difference between these two is that for the Markov switching models, our only estimate of the observed information matrix uses second order finite differences to compute the Hessian, whereas in state space models we also have a version computed using only first order finite differences.

For both, `robust` means a sandwich covariance estimator, with the outer product of gradients in the middle, but state space models have two options: `robust_oim` for the first-order-differences-based Hessian and `robust_approx` for the second-order-differences-based method, where the default is `robust`, and that maps to `robust_oim`.

Does that make sense? Other than the bug you found, are there more changes we should consider?

**Comment by josef-pkt** [comment]
> `robust_oim` for the first-order-differences-based Hessian

If this is based on first derivatives, then shouldn't it be OPG, i.e. `robust_opg`. Or do you mean something different?

**Comment by ChadFulton** [comment]
> If this is based on first derivatives, then shouldn't it be OPG, i.e. robust_opg. Or do you mean something different?

In Harvey's book (section 3.4.5), he shows that the expected information matrix written in terms of the Hessian can be computed using only first derivative terms, and then the observed information matrix just drops the expectations. But it's not a OPG computation, it's still a computation of the Hessian-based estimate of the information matrix.

**Comment by josef-pkt** [comment]
@ChadFulton Thank, now it sounds familiar, you might have answered this already once.  
(I went over some sandwich sections in Wooldridge's textbook yesterday, and I always find it confusing how many different asymptotically equivalent versions there are for this.)

**Comment by jbrockmendel** [comment]
> Other than the bug you found, are there more changes we should consider?

That was the most important one by a wide margin.  Next up is just a matter of aliasing:

`MLEResults.cov_params_robust` is an alias for `cov_params_robust_oim`.  AFAICT, `MarkovSwitchingResults.cov_params_robust` could be renamed to `cov_params_robust_oim` and have an identical aliasing structure.  No idea whether the Harvey (1989) reference in the mlemodel docstring is applicable in the markov_switching case.

The last name question is if there is something in `markov_switching` that could reasonably be labelled `_cov_params_robust_approx`.  I think that might need to be a new method, and it is pretty clear what that method would look like, but I don't know if there was an econometric reason for omitting that in the first place.

Finally on a simplifying-code (and slightly improving coverage) note, the following shows up in `MLEResults._cov_params_approx` and `_cov_params_robust_approx`

```
        if approx_complex_step:
            evaluated_hessian = self.model._hessian_complex_step(
                self.params, transformed=True
                )
        else:
            evaluated_hessian = self.model._hessian_finite_difference(
                self.params, transformed=True,
                approx_centered=approx_centered
                )
```

But that could be simplified to one line:

```
evaluated_hessian = self.model.hessian(self.params,
    transformed=True,
    method='approx', 
    approx_complex_step=approx_complex_step,
    approx_centered=approx_centered
    )
```

This has the side-benefit of getting coverage for the `method == 'approx' and approx_complex_step` case in `hessian`.  Also I don't think the `not approx_complex_step` case is currently hit in either `_cov_params_approx` or `_cov_params_robust_approx`.

Finally, it looks like in `_get_robustcov_results` methods, the `res.cov_kwds['descriptions']` are almost identical, but are a bit more specific in mlemodel.  (Notwithstanding the name "oim" which is not yet specified in markov_switching).

**Comment by jbrockmendel** [comment]
@ChadFulton 

>>In mlemodel, if cov_type="robust_approx", it gets passed through to cov_params_robust, which I find counter-intuitive given that there exists a cov_params_robust_approx method.

> This looks like a bug to me - it looks like there's no way to get the cov_params_robust_approx. Thanks for pointing that out.

Ready for a PR to fix this?  It's a super-simple fix, but I'll need to defer to you on following up with any needed test changes/improvements.

**Comment by jbrockmendel** [comment]
The markov_switching descriptions are epsilon way from matching the mlemodel descriptions.  e.g.


```
        elif self.cov_type == 'opg':
            res.cov_params_default = res.cov_params_opg
            res.cov_kwds['description'] = (
                'Covariance matrix calculated using the outer product of'
                ' gradients.'
            )

vs.

        elif self.cov_type == 'opg':
            res.cov_params_default = res.cov_params_opg
            res.cov_kwds['description'] = (
                'Covariance matrix calculated using the outer product of'
                ' gradients (%s).' % approx_type_str
            )
```

Is there an appropriate `approx_type_str` we could add to the descriptions in markov_switching to make these identical?

**Comment by ChadFulton** [comment]
> Ready for a PR to fix this? It's a super-simple fix, but I'll need to defer to you on following up with any needed test changes/improvements.

I think all of the things we discussed in this thread are good and straightforward to check, I can commit to reviewing something this week.

> Is there an appropriate approx_type_str we could add to the descriptions in markov_switching to make these identical?

That's a nice improvement to consistency; it looks like the Markov switching Hessian is always computed using complex step, so we could put:

```python
        elif self.cov_type == 'opg':
            res.cov_params_default = res.cov_params_opg
            res.cov_kwds['description'] = (
                'Covariance matrix calculated using the outer product of'
                ' gradients (complex-step).'
            )
```

**Comment by jbrockmendel** [comment]
Closed by #4056+#4029
